core/bin/dust_api.rs

```diff
@@ -1156,6 +1157,126 @@ async fn data_sources_search(
     }
 }
 
+/// Update tags of a document in a data source.
+
+#[derive(serde::Deserialize)]
+struct DataSourcesDocumentsUpdateTagsPayload {
+    add_tags: Option<Vec<String>>,
+    remove_tags: Option<Vec<String>>,
+}
+
+async fn data_sources_documents_update_tags(
+    extract::Path((project_id, data_source_id, document_id)): extract::Path<(i64, String, String)>,
+    extract::Json(payload): extract::Json<DataSourcesDocumentsUpdateTagsPayload>,
+    extract::Extension(state): extract::Extension<Arc<APIState>>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+    let add_tags = match payload.add_tags {
+        Some(tags) => tags,
+        None => vec![],
+    };
+    let remove_tags = match payload.remove_tags {
+        Some(tags) => tags,
+        None => vec![],
+    };
+    let add_tags_set: HashSet<String> = add_tags.iter().cloned().collect();
+    if add_tags.len() != add_tags_set.len() {
```

Comment:
not sure these errors are really needed. I would remove them

Reply:
Ok for me. Should I still dedup though ?

Reply:
kept the de-dup (rust's `.collect()` is awesome)

Reply:
yes yes dedup but not fail :+1: